### ISSUE TITLE
Support for chanjo2

### ISSUE THREAD

**Issue by northwestwitch** [Clinical-Genomics/scout#4000]
Add the option to use [chanjo2](https://chanjo2-stage.scilifelab.se/docs) to get coverage stats.

For now the chanjo2 endpoints are open, but we will probably restrict the access to most of them when it goes in  production. 

We should also replace chanjo-report with a new solution (in scout? In chanjo2?)

**Comment by Jakob37** [comment]
Hello from Lund! 

Can I ask what the status on this issue is?

We have been talking about using Chanjo2 for coverage calculations. We had to stop Chanjo1 due to being too heavy on our heavily-loaded Scout environment. We were thinking Chanjo2 would work better with its better performance (?). But if still far off we'll need to think about how to approach it.

**Comment by dnil** [comment]
@northwestwitch will be most up to date when she is back on track. The situation is basically that the core functionality is in place in chanjo2, but we (collectively if you are interested) next need to replace or refactor the reporting functionality from chanjo-report so we can produce neat reports from chanjo2 in or from scout. There is also a bit of work left on syncing up all reference resources between pipelines, chanjo2 and scout (see the shug repo), but it is well underway.

**Comment by Jakob37** [comment]
Thanks for the update! That is good to know. We would be happy to help out. Are there sub-parts of this that you think would suit well for us to look at?

**Comment by northwestwitch** [comment]
Hi both! As Daniel mentioned before, the functionally of chanjo2 is already in place, including the possibility of creating both coverage report ([demo](https://chanjo2-stage.scilifelab.se/report/demo)) and gene coverage overview ([demo](https://chanjo2-stage.scilifelab.se/overview/demo)) - the 2 report types that Scout currently collects from chanjo.

What needs to be done is making Scout aware of the possibility to use chanjo2, codewise

The fact that you'd like to use chanjo2, @Jakob37, is a great opportunity for us, since we are not quite there yet with the creation of the necessary files from the pipeline in our facility and we haven't had the possibility to test chanjo2 with real data yet.

So basically what needs t be done at the end of the pipeline is creating [d4 files](https://github.com/38/d4-format) out of the BAM/CRAM files, one for each sample of the analysis. 

In the meanwhile we can work on fixing Scout to include the path to the d4 file for each case sample in the yaml config file used when loading cases to Scout. This, a few fixes to the templates with the links to chanjo2 and a new config setting pointing to a running instance of chanjo2 should be sufficient to have the chanjo2 reporting in place!

**Comment by Jakob37** [comment]
That sounds really good! I'll start working on producing d4 files from our pipeline right away then.

I am on parental leave every second week at the moment, but expect to have this as my highest prio the week after next.

**Comment by northwestwitch** [comment]
> That sounds really good! I'll start working on producing d4 files from our pipeline right away then.
> 
> I am on parental leave every second week at the moment, but expect to have this as my highest prio the week after next.

Cool, let's fix it! :muscle:

**Comment by Jakob37** [comment]
Hi again! I have made a branch of our wgs pipeline now which produces one d4 file per sample. Should be ready for testing with Scout when that is ready :)

**Comment by northwestwitch** [comment]
> Hi again! I have made a branch of our wgs pipeline now which produces one d4 file per sample. Should be ready for testing with Scout when that is ready :)

That's' great, I am working on it  right now. I'll let you know when we are ready with a PR. I think I'm going to split it into PRs actually, one taking care of saving the path to the d4 files, and one running queries to chanjo2.

Perhaps you try to launch chanjo2 as a service on your machine? So it will be available when also scout is ready

**Comment by Jakob37** [comment]
> That's' great, I am working on it right now. I'll let you know when we are ready with a PR. I think I'm going to split it into 2 PRs actually, one taking care of saving the path to the d4 files, and one running queries to chanjo2.
> 
> Perhaps you could try to launch chanjo2 as a service on your machine? So it will be available when also scout is ready

OK, sounds good! Yes, that sounds like a good idea, I'll start looking into getting chanjo2 running.

**Comment by northwestwitch** [comment]
> Hello again! Just wanted to say we have Chanjo2 up and running locally, and it works querying it for info from a d4-file. We are ready to continue testing when the other parts are in place, just let us know!

That's great. I'm working on integrating chanjo2 into scout now. I hope I'll be having a PR ready this week!

**Comment by Jakob37** [comment]
> > Hello again! Just wanted to say we have Chanjo2 up and running locally, and it works querying it for info from a d4-file. We are ready to continue testing when the other parts are in place, just let us know!
> 
> That's great. I'm working on integrating chanjo2 into scout now. I hope I'll be having a PR ready this week!

Sounds great, no rush!

**Comment by Jakob37** [comment]
> I've started to test the scout-chanjo2 integration then I've realised that current chanjo2 is way too slow to return responses when you have gene panels with thousands of genes: It takes minutes to compute coverage and coverage completeness for a range of coverage thresholds on the fly, with the result that the page times out.
> 
> I guess I should have tested with prod-like real files before but it didn't feel urgent until you said you wanted chanjo2 in prod @Jakob37 (which is super-good by the way! 😆 ).
> 
> I basically I found out that:
> 
>     * **The pyd4 - python library wrapper to d4tools (rust language) - is very slow** and can't be used when you have this amount of computation to run
> 
>     * **You need multiprocessing** in order to speed up the whole thing
> 
> 
> So I started to re-write chanjo2 accordingly, see [Clinical-Genomics/chanjo2#223](https://github.com/Clinical-Genomics/chanjo2/issues/223) and [Clinical-Genomics/chanjo2#224](https://github.com/Clinical-Genomics/chanjo2/pull/224)

Thank you for the update @northwestwitch ! And great that you are working on this. Let us know if/when we can help out.

**Comment by northwestwitch** [comment]
Ok I finally have a [workable PR for chanjo2](https://github.com/Clinical-Genomics/chanjo2/pull/236) that doesn't time out when producing reports with many thousands of genes. I have used it it a combo with [this scout PR](https://github.com/Clinical-Genomics/scout/pull/4402) and looks like it's working. If you want you could give it a try @Jakob37 . I have tried to write clear instructions on how to test on the scout PR but let me know if I missed anything when describing the test setup

**Comment by Jakob37** [comment]
> Ok I finally have a [workable PR for chanjo2](https://github.com/Clinical-Genomics/chanjo2/pull/236) that doesn't time out when producing reports with many thousands of genes. I have used it it a combo with [this scout PR](https://github.com/Clinical-Genomics/scout/pull/4402) and looks like it's working. If you want you could give it a try @Jakob37 . I have tried to write clear instructions on how to test on the scout PR but let me know if I missed anything when describing the test setup

Great to hear! I'll start testing it early this week

**Comment by Jakob37** [comment]
Hi again! Just another follow up. I discussed the Chanjo2 progress with our geneticists, and they were really happy to hear that things are progressing :)

They asked for a different piece of functionality related to coverage, which seems to not currently be running with Chanjo2. They said they often checks coverage for exons in a single gene.

In production Scout there are coverage links for individual genes like this (I just added a gene to test/demonstrate):

<img width="345" alt="chanjo2_single_gene_coverage" src="https://github.com/Clinical-Genomics/scout/assets/7374399/353931d9-0f2c-4c8e-909f-ddc4e32b38b0">

While for the Scout+Chanjo2 version I tested out, it does not appear:

<img width="345" alt="chanjo2_single_gene_wo_coverage" src="https://github.com/Clinical-Genomics/scout/assets/7374399/3be7f219-f8e6-4517-8386-32d49f87e578">

Have I understood / tested this correctly, is this something that also will be updated for Chanjo2?

Meanwhile I am working on putting the d4 file generation into our production pipeline, and preparing the Scout/Chanjo2 setup for being deployed in our Scout production. Getting close now!

**Comment by northwestwitch** [comment]
Hi @Jakob37, great to see that the clinicians appreciated that we are working on this! I'm currently working on this.
So far we have introduced the chanjo2 coverage for other gene panels (see #4518). Chanjo2 coverage for HPO panels was the next PR schedule for today.

The report where you see uncovered genes is the coverage overview and I'm also going to add links to that over that in the same places over the next days. 

I might ask you for help testing a few of these functionality if you have time? I'll keep you posted!

**Comment by Jakob37** [comment]
> Hi @Jakob37, great to see that the clinicians appreciated that we are working on this! It is actually a work in progress, but what you ask is coming. So far we have introduced the chanjo2 coverage for other gene panels (see #4518). Chanjo2 coverage for HPO panels was the next PR scheduled for today.
> 
> The report where you see uncovered genes is the coverage overview and I'm also going to add links to that over the next days.
> 
> I might ask you for help testing a few of these functionality if you have time? I'll keep you posted!

OK good to hear! Thanks for the clarification.

I am definitely happy to test. Just ping me. I am working full time again now, so available all (most) days :)
